lib/browser/api/navigation-controller.js

```diff
@@ -55,9 +55,35 @@ const NavigationController = (function () {
     if (options == null) {
       options = {}
     }
+    const p = new Promise((resolve, reject) => {
+      const finishListener = () => {
+        this.webContents.removeListener('did-fail-load', failListener)
+        resolve()
+      }
+      const failListener = (event, errorCode, errorDescription, validatedURL, isMainFrame, frameProcessId, frameRoutingId) => {
+        if (!isMainFrame) {
+          return
+        }
+        this.webContents.removeListener('did-finish-load', finishListener)
+        const err = new Error(`${errorDescription} (${errorCode}) loading '${validatedURL}'`)
+        Object.assign(err, {
+          errno: errorCode,
+          code: errorDescription,
+          validatedURL,
+          frameProcessId,
+          frameRoutingId
+        })
+        reject(err)
+      }
+      this.webContents.once('did-finish-load', finishListener)
```

Comment:
I know this is getting picky but if we're exposing this as API we should think about it.

Is it possible for this `did-finish-load` event be for a previous load?  E.g. if two calls are sequentially made to `loadURL` will the first promise resolve or reject?  If a `loadURL` call is made a few ticks before a different navigation finishes would this promise be resolved eroneously?

Unless we validate a navigation request ID or something I don't think we can 100% sure this is the same navigation

Reply:
Yeah... it's also possible for the promise to _never resolve_, which I noticed when trying to test load failures with a bad scheme (like `bad-scheme://foo`). The navigation gets cancelled and you're left hanging.

I think these are fundamental issues with the API as it stands (the promise is just exposing the existing flaws) and we should probably see if we can fix them.

Reply:
fixed 'em